diffprivlib/models/tree.py

```diff
@@ -0,0 +1,602 @@
+
+        if random_state is not None:
+            np.random.seed(random_state)
+
+        warn_unused_args(unused_args)
+
+    @property
+    def n_features(self):
+        return self._n_features
+
+    @property
+    def n_classes(self):
+        return self._n_classes
+
+    @property
+    def classes_(self):
+        return self._classes
+
+    @property
+    def cat_features(self):
+        return self._cat_features
+
+    @property
+    def max_depth(self):
+        return self._max_depth
+
+    @property
+    def estimators(self):
+        return self._estimators
+
+    @property
+    def feature_domains(self):
+        return self._feature_domains
+
+    def fit(self, X, y):
+        """Fit the model to the given training data.
+
+        Parameters:
+            X : array-like, shape (n_samples, n_features)
+                Training vector, where n_samples is the number of samples and n_features is the number of features.
+
+            y : array-like, shape (n_samples,)
+                Target vector relative to X.
+
+        Returns:
+            self: class object
+        """
+        if not isinstance(self._n_estimators, numbers.Integral) or self._n_estimators < 0:
+            raise ValueError(f'Number of estimators should be a positive integer; got {self._n_estimators}')
+
+        if not isinstance(self._cat_feature_threshold, numbers.Integral) or self._cat_feature_threshold < 0:
+            raise ValueError('Categorical feature threshold should be a positive integer;'
+                             f'got {self._cat_feature_threshold}')
+
+        self._accountant.check(self._epsilon, 0)
+
+        X, y = np.array(X), np.array(y)
+        check_X_y(X, y)
+
+        self._n_features = X.shape[1]
+        self._cat_features = get_cat_features(X, self._cat_feature_threshold)
+        self._max_depth = calc_tree_depth(n_cont_features=self._n_features-len(self._cat_features),
+                                          n_cat_features=len(self._cat_features), max_depth=self._max_depth)
+        self._classes = np.unique(y)
+        self._n_classes = len(self._classes)
+
+        if self._feature_domains is None:
+            warnings.warn(
+                "`feature_domains` parameter hasn't been specified, so falling back to determining domains from the data.\n"
+                "This may result in additional privacy leakage. To ensure differential privacy with no "
+                "additional privacy loss, specify `feature_domains` according to the documentation",
+                PrivacyLeakWarning)
+            self._feature_domains = get_feature_domains(X, self._cat_features)
+
+        if len(self._feature_domains) != self._n_features:
+            raise ValueError("Missing domains for some features in `feature_domains`")
+
+        if self._n_estimators > len(X):
+            raise ValueError('Number of estimators is more than the available samples')
+
+        subset_size = int(len(X) / self._n_estimators)
```

Comment:
General comment here: If our dataset has 99 entries, and we have 10 estimators, this will result in 9 estimators with 9 entries, and the last estimator with 18 entries? Perhaps there's a better way to do this, but not crucial for a first draft to be merged.

Reply:
I think this line is a holdover from my original repo. The intent is for each tree to have a roughly-equal portion of the data. The subsets need to be disjoint (no overlap) in order for the privacy guarantees to hold.